File: frontend/src/geo.ts
/**
 * Ray-casting point-in-polygon over a boundary of [lat, lng] vertices (the
 * closing edge back to the first point is implied). Treats longitude as x and
 * latitude as y — fine at festival scale, where the flat-earth approximation
 * costs nothing. Shared by both MapScreen variants.
 */
export function pointInPolygon(
  point: { latitude: number; longitude: number },
  polygon: [number, number][]
): boolean {
  const { latitude: y, longitude: x } = point;
  let inside = false;
  for (let i = 0, j = polygon.length - 1; i < polygon.length; j = i++) {
    const yi = polygon[i][0];
    const xi = polygon[i][1];
    const yj = polygon[j][0];
    const xj = polygon[j][1];
    const intersects = yi > y !== yj > y && x < ((xj - xi) * (y - yi)) / (yj - yi) + xi;
    if (intersects) inside = !inside;
  }
  return inside;
}

/** Landmarks whose geofence contains the given point (empty if point is null). */
export function landmarksContaining<T extends { boundary: [number, number][] | null }>(
  landmarks: T[],
  point: { latitude: number; longitude: number } | null
): T[] {
  if (!point) return [];
  return landmarks.filter(
    (l) => l.boundary && l.boundary.length >= 3 && pointInPolygon(point, l.boundary)
  );
}

File: backend/app/geo.py
"""Point-in-polygon over the [lat, lng] boundaries stored in `events.boundary`
and `landmarks.boundary`.

A deliberate port of `pointInPolygon` in `frontend/src/geo.ts` — the same
ray-casting loop with the same edge behaviour, so the client's "you're at this
stage" pill and the server's attendance credit can never disagree about a
position sitting on a boundary edge.

This is plain Python over JSON columns the row already carries: no spatial
extension, no geometry type, no spatial index or query. "PostGIS/spatial
queries" stays out of scope.
"""

# A boundary needs at least a triangle to enclose anything. Mirrors the
# `boundary.length >= 3` guard in landmarksContaining().
MIN_POLYGON_POINTS = 3


def point_in_polygon(lat: float, lng: float, polygon: list | None) -> bool:
    """True if (lat, lng) falls inside the polygon of [lat, lng] vertices.

    The closing edge back to the first vertex is implied. Longitude is treated
    as x and latitude as y — the flat-earth approximation costs nothing at
    festival scale. Anything degenerate (None, too few points, a malformed
    vertex) is simply "not inside" rather than an error: boundaries are
    admin-authored JSON, and a bad one must not break the location hot path.
    """
    if not polygon or len(polygon) < MIN_POLYGON_POINTS:
        return False
    inside = False
    count = len(polygon)
    j = count - 1
    for i in range(count):
        try:
            yi, xi = float(polygon[i][0]), float(polygon[i][1])
            yj, xj = float(polygon[j][0]), float(polygon[j][1])
        except (TypeError, ValueError, IndexError, KeyError):
            return False
        if (yi > lat) != (yj > lat) and lng < (xj - xi) * (lat - yi) / (yj - yi) + xi:
            inside = not inside
        j = i
    return inside
